ckcp/hack/util/reset_ckcp.sh

```diff
@@ -0,0 +1,288 @@
+        Default value: \$KUBECONFIG
+    -d, --debug
+        Activate tracing/debug mode.
+    -h, --help
+        Display this message.
+
+Example:
+    %s --kcp-kubeconfig './tmp/tmp.435kjkdsf/credentials/kubeconfig/kcp/ckcp.kubeconfig'
+" "${0##*/}" "${0##*/}" >&2
+}
+
+parse_args() {
+  while [[ $# -gt 0 ]]; do
+    case "$1" in
+    -k | --kubeconfig)
+      shift
+      KUBECONFIG="$1"
+      ;;
+    --kcp-kubeconfig)
+      shift
+      KCP_KUBECONFIG="$1"
+      ;;
+    --reset-hard)
+      shift
+      RESET_HARD="$1"
+      ;;
+    -d | --debug)
+      set -x
+      ;;
+    -h | --help)
+      usage
+      exit 0
+      ;;
+    --)
+      # End of arguments
+      break
+      ;;
+    *)
+      exit_error "Unknown argument: $1"
+      ;;
+    esac
+    shift
+  done
+}
+
+prechecks() {
+  KCP_KUBECONFIG=${KCP_KUBECONFIG:-}
+  if [[ -z "${KCP_KUBECONFIG}" ]]; then
+    printf "\n\n[ERROR] Missing parameter --kcp-kubeconfig"
+    exit
+  fi
+}
+
+# Removes different Argo CD applications deployed
+uninstall_pipeline_service(){
+    printf "\n  Uninstalling Pipeline Service:\n"
+    # Remove Pipeline-Service Argo CD application
+    kubectl delete -f "$GITOPS_DIR/argocd/argocd.yaml"
+    # Remove all the child applications deployed by Pipeline-Service
+    kubectl delete -k "$GITOPS_DIR/argocd/argo-apps"
+
+    # Check if the Argo CD applications have been indeed removed
+    # list of all the Argo Apps that a user can deploy using Pipeline Service
+    mapfile -t all_argo_apps < <(kubectl kustomize "$GITOPS_DIR/argocd/argo-apps"  | yq '.metadata.name' | grep -e '[a-z]')
+    all_argo_apps+=("pipeline-service")
+    # list of all the Argo Apps still deployed
+    mapfile -t argo_apps_deployed < <(argocd app list -o yaml | yq '.[].metadata.name')
+    matched_apps=()
+    for app in "${all_argo_apps[@]}"
+    do
+        for deployed_app in "${argo_apps_deployed[@]}"
+        do
+            if [ "$app" == "$deployed_app" ];
+            then
+              matched_apps+=("$app")
+            fi
+        done
+    done
+    if (( ${#matched_apps[@]} >= 1 )); then
+        printf "\n\n[ERROR] Couldn't uninstall deployed Argo CD applications:%s ${matched_apps[*]}"
+        exit
+    fi
+    printf "\nAll the Argo CD applications are successfully uninstalled.\n"
+}
+
+uninstall_operators(){
+    printf "\n  Uninstalling Openshift-GitOps Operator:\n"
+    kubectl delete -k "$CKCP_DIR/openshift-operators/openshift-gitops"
+    openshift_gitops_csv=$(kubectl get csv -n openshift-operators | grep -ie "openshift-gitops-operator" | cut -d " " -f 1)
+    if [[ -n "$openshift_gitops_csv" ]];
+    then
+      kubectl delete csv -n openshift-operators "$openshift_gitops_csv"
+    fi
+    mapfile -t argo_crds < <(kubectl get crd | grep -iE "argoproj.io|gitopsservices" | cut -d " " -f 1)
+    if [[ "${#argo_crds[@]}" -gt 0 ]];
+    then
+      for crd in "${argo_crds[@]}"
+      do
+        kubectl delete crd "$crd" &
```

Comment:
Why is it run in the background? If you patched the resource first to remove the finalizer, and then deleted it, shouldn't it work in the foreground?

If this requires a change, this appears multiple times in the file.

Reply:
As discussed, the patch operation before deleting(i.e., resource is in terminating state) is not allowed. So, we'll keep it in favor of removing(forcefully) CRDs.